File: src/go/problems/MaxDepth.go
// https://leetcode.com/problems/maximum-depth-of-binary-tree/

package problems

import (
	"testing"
)

func TestMaxDepth(t *testing.T) {
	var node = TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &ListNode{Val: 3}}
	
	result := MaxDepth(node)
	t.Log(result)
}

func MaxDepth(root *TreeNode) int {
	return Dfs(root, 0)
}

func Dfs(node *TreeNode, depth int) int {
	if node == nil {
		return depth
	}

	left := Dfs(node.Left, depth)
	right := Dfs(node.Right, depth)

	return 1 + max(left, right)
}

func max(left int, right int) int {
	if left > right {
		return left
	}

	return right
}

File: src/python/problems/max_depth.py
# https://leetcode.com/problems/maximum-depth-of-binary-tree/
from typing import Optional
from models.tree_node import TreeNode


class MaxDepth:
    def Solution(self, root: Optional[TreeNode]) -> int:
        def Dfs(node, depth) -> int:
            if not node:
                return depth
            
            left = Dfs(node.left, depth)
            right = Dfs(node.right, depth)

            return 1 + max(left, right)

        return Dfs(root, 0)


MaxDepth().Solution(TreeNode(1, TreeNode(2), TreeNode(3)))
